### DISCUSSION TITLE
Detecting connection changes

### DISCUSSION THREAD

**Discussion by hognevevle** [enisdenjo/graphql-sse#77] [Q&A] [answered]
Hi there!

Is there a way to detect when a (re)connection and connection loss has occured? I've tried `onMessage`, but it doesn't seem to give me anything there.

**Comment by enisdenjo** [comment]
Not really, because each connection ([in "distinct connection mode"](https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md#distinct-connections-mode), default) is a separate connection and its therefore hard to provide a centralised dispatcher of this event.

You could theoretically used the [`retry` callback](https://the-guild.dev/graphql/sse/docs/interfaces/client.ClientOptions#retry) which controls the reconnections.

Out of curiosity, what are you trying to achieve? What is the use-case?

**Reply by hognevevle** [reply]
Thanks for confirming @enisdenjo .

I understand how it could be challenging.. 

I don't think `retry` will let me do what I need, since I have no easy way of knowing if the retry will be successful or not, I'll just know that something intends to retry.

We're using graphql-sse for subscriptions, and "regular" http for queries/mutations. Our use case is that whenever a client's connection is lost and then reestablished, we want to refetch all active queries to make sure the client's state is updated (because subscription messages sent via SSE were likely not delivered while the connection was gone). 🤔

**Reply by hognevevle** [reply]
@enisdenjo What about the keepalive-messages sent from the server - would we be able to monitor these somehow? If so, I'd imagine we could detect connection loss if we don't see any keepalives for some given time. Then, if we start seeing them again, we'll know connection is back up and we can make sure to refetch all the (potentially) missed data.

As far as I can see, the client discards keepalive-messages by default. Is there a way they can be exposed instead?

**Reply by hognevevle** [reply]
Or perhaps adding a callback that fires every time the subscription is established? 

![CleanShot 2023-11-03 at 11 43 16](https://github.com/enisdenjo/graphql-sse/assets/2721188/b38a1252-4f8a-4db9-ba7a-c4b6cc6d2914)

**Reply by enisdenjo** [reply]
Hey sorry for the late reply. I am a tad busy at the moment - but I haven't forgotten about this! Will revisit after my time frees up. Thanks for understanding!

**Comment by enisdenjo** [comment, marked as answer]
Hey @hognevevle, I'll start by apologising for such a long wait. Nevertheless, I bear great news! With [v2.4.0](https://github.com/enisdenjo/graphql-sse/releases/tag/v2.4.0), we have more sophisticated event listeners (including what was discussed here).

Please consult the description of #84 for explanation with examples.

**Reply by hognevevle** [reply]
This is really awesome, thank you so much! 🙏🥳

**Reply by enisdenjo** [reply]
You're very welcome! Would appreciate it if you can try it out and report back whether this works for you as intended. Thanks!
